text/0000-const-generics.md

```diff
@@ -0,0 +1,287 @@
+fully evaluated at compile time. Unlike statics, consts will be inlined at
+their use sites rather than existing in the data section of the compiled
+binary.
+
+* __Const parameter (generic const):__ A const which a type or function is
+abstract over; this const is input to the concrete type of the item, such as
+the length parameter of a static array.
+
+* __Associated const:__ A const associated with a trait, similar to an
+associated type. Unlike a const parameter, an associated const is *determined*
+by a type.
+
+* __Const variable:__ Either a const parameter or a trait, contrast with
+concrete const; a const which is undetermined in this context (prior to
+monomorphization).
+
+* __Concrete const:__ In contrast to a const variable, a const which has a
+known and singular value in this context.
+
+* __Const expression:__ An expression which evaluates to a const. This may be
+an identity expression or a more complex expression, so long as it can be
+evaluated by Rust's const system.
+
+* __Abstract const expression:__ A const expression which involves a const
+variable (and therefore the value that it evaluates to cannot be determined
+until after monomorphization).
+
+* __Const projection:__ The value of an abstract const expression (which cannot
+be determined in a generic context because it is dependent on a const
+variable).
+
+* __Identity expression:__ An expression which cannot be evaluated further
+except by substituting it with names in scope. This includes all literals as
+well all idents - e.g. `3`, `"Hello, world"`, `foo_bar`.
+
+## Declaring a const parameter
+
+In any sequence of type parameter declarations (such as in the definition of a
+type or on the `impl` header of an impl block) const parameters can also be
+declared. Const parameters always come after type parameters, and their
+declarations take the form `const $ident: $ty`, as in:
+
+```rust
+struct RectangularArray<T, const WIDTH: usize, const HEIGHT: usize> {
+    array: [[T; WIDTH]; HEIGHT],
+}
+```
+
+The idents declared are the names used for these const parameters
+(interchangeably called "const variables" in this RFC text), and all values
+must be of the type ascribed to it. Which types can be ascribed to const
+parameters is restricted later in this RFC.
+
+The const parameter is in scope for the entire body of the item (type, impl,
+function, method, etc) in which it is declared.
+
+## Applying a const as a parameter
+
+Any const expression of the type ascribed to a const parameter can be applied
+as that parameter. When applying an expression as const parameter (except for
+arrays), which is not an identity expression, the expression must be contained
+within a block. This syntactic restriction is necessary to avoid requiring
+infinite lookahead when parsing an expression inside of a type.
+
+```rust
+const X: usize = 7;
+
+let x: RectangularArray<i32, 2, 4>;
+let y: RectangularArray<i32, X, {2 * 2}>;
```

Comment:
>Am I remembering correctly that `struct Foo;` puts `Foo` in both namespaces?

That't true. Also `struct Foo(u8, u8);`.
So, the namespace separation is used all the time and cannot be deprecated, this is misunderstanding from the previous commenters.